src/retime.jl

```diff
@@ -0,0 +1,233 @@
+_toAggregationMethod(::Val{:sum}) = Sum()
+_toAggregationMethod(::Val{:median}) = Median()
+_toAggregationMethod(::Val{:first}) = First()
+_toAggregationMethod(::Val{:last}) = Last()
+_toAggregationMethod(x::AggregationMethod) = x
+
+_toExtrapolationMethod(x::Symbol) = _toExtrapolationMethod(Val(x))
+_toExtrapolationMethod(::Val{:fillconstant}) = FillConstant(0.0)
+_toExtrapolationMethod(::Val{:nearest}) = NearestExtrapolate()
+_toExtrapolationMethod(::Val{:missing}) = MissingExtrapolate()
+_toExtrapolationMethod(::Val{:nan}) = NaNExtrapolate()
+_toExtrapolationMethod(x::ExtrapolationMethod) = x
+
+function retime(ta, new_dt::Dates.Period; kwargs...)
+    new_timestamps = timestamp(ta)[1]:new_dt:timestamp(ta)[end]
+    return retime(ta, new_timestamps; kwargs...)
+end
+
+function retime(ta, period::Function; kwargs...)
+    new_timestamps = map(i -> first(i), _split(timestamp(ta), period))
+    return retime(ta, new_timestamps; kwargs...)
+end
+
+function retime(
+    ta::TimeArray{T,N,D,A},
+    new_timestamps::AbstractVector{DN};
+    upsample::Union{Symbol,InterpolationMethod}=Previous(),
+    downsample::Union{Symbol,AggregationMethod}=Mean(),
+    extrapolate::Union{Symbol,ExtrapolationMethod}=NearestExtrapolate(),
+    skip_missing::Bool=true,
+) where {T,N,D,A,DN}
+    upsample = _toInterpolationMethod(upsample)
+    downsample = _toAggregationMethod(downsample)
+    extrapolate = _toExtrapolationMethod(extrapolate)
+
+    new_values = __get_new_values(
+        T, length(new_timestamps), size(values(ta), 2), extrapolate, skip_missing
+    )
+    old_timestamps = convert(Vector{DN}, timestamp(ta))
+    old_values = values(ta)
+    @views begin
+        for col_i in 1:size(old_values, 2)
+            if skip_missing
+                idx = findall(x -> !ismissing(x) && !isnan(x), old_values[:, col_i])
+            else
+                idx = ones(Int, length(old_timestamps))
+            end
+
+            _retime!(
+                new_values[:, col_i],
+                old_timestamps[idx],
+                old_values[idx, col_i],
+                new_timestamps,
+                upsample,
+                downsample,
+                extrapolate,
+                skip_missing,
+            )
+        end
+    end
+    return TimeArray(new_timestamps, new_values, colnames(ta), meta(ta))
+end
+
+function _retime!(
+    new_values::AbstractVector{AN},
+    old_timestamps::AbstractVector{D},
+    old_values::AbstractVector{A},
+    new_timestamps::AbstractVector{D},
+    upsample::InterpolationMethod,
+    downsample::AggregationMethod,
+    extrapolate::ExtrapolationMethod,
+    skip_missing::Bool,
+) where {D,AN,A}
+    x = Dates.value.(old_timestamps)
+    x_min, x_max = extrema(x)
+    x_new = Dates.value.(new_timestamps)
+
+    N = length(x_new)
+
+    @views begin
+        # check each interval between i and i+1 if there is no or one sample (upsample), more than one sample (downsample)
+        for i in 1:N
```

Comment:
IIUC, this loop can be rewritten as :

```julia
for i in 1:N 
    new_values[i] =
        if x_new[i] < x_min || x_new[i] > x_max
            ...
        else
            ...
        end
end
```

Then,  I think the main algorithm could be reconstructed as two step
step 1. is something like `groupby` the new timestamps then figure out the proper transformation function. And let's call this is the planning step.
step 2. is for filling the actual value.

I think the result of planning step could be kept/shared/cached for the case of `skip_missing=false` and make the performance better.

Reply:
Just slapping `Threads.@threads` to both for-loops is actually worse, also see: https://discourse.julialang.org/t/editing-an-array-with-multiple-threads/25364/12

```julia
using MarketData, TimeSeries, Dates
new_timestamps = collect(Dates.Date(2000):Dates.Week(1):Dates.Date(2001))
```

With `Threads.@threads`
```julia
julia> @btime retime($ohlc, $new_timestamps);
  16.400 μs (459 allocations: 93.64 KiB)
```

without:
```julia
julia> @btime retime($ohlc, $new_timestamps);
  9.600 μs (49 allocations: 36.45 KiB)
```

Reply:
If performance becomes an issue, we can get back to this.